src/dataflow/src/render/mod.rs

```diff
@@ -572,3 +598,42 @@ where
         }
     }
 }
+
+use differential_dataflow::lattice::Lattice;
+use timely::progress::timestamp::Refines;
+
+/// A timestamp type that can be used for operations within MZ's dataflow layer.
+pub trait RenderTimestamp: Timestamp + Lattice + Refines<mz_repr::Timestamp> {
```

Comment:
How is `Refines` useful (in general, not just here) if it's unable to express failures when converting between two types? It seems that we couldn't define this between u64 and i64, for example.

Reply:
It is not used for that. It is used to express e.g. that `(T, usize)` refines `T`. Think of it as meaning "represents a subdivision of the other time".